### ISSUE TITLE
[RFC] Upstream current implementation of ssd_offload from fairscale FSDP to Torch Distributed FSDP

### ISSUE THREAD

**Issue by another-pjohnson** [pytorch/pytorch#76924]
### 🚀 The feature, motivation and pitch

# Goal
Upstream the currently experimental ssd_offload feature from Fairscale FSDP to torch distributed FSDP.

ssd_offload is a capability in similar FSDP-like systems (for example zero-infinity by Microsoft)

# API changes:
Logically from the outside ssd_offload looks very similar to cpu_offload. So merging the options probably makes sense (even though in the implementation they will look quite different.

fsdp constructor: cpu_offload argument -> offload_config
torch.distributed.fsdp.CPUOffload -> ... OffloadConfig

```
class OffloadConfig:
    offload_type: Optional(str) = None # "cpu" or "ssd" only supported for now, maybe make into an enum
    offload_directory: Optional(str) = None # directory where tensors offloaded to disk reside (only valid with offload_type="ssd". In the future this will probably be extended to support some sort of wildcards (e.g. a single machine has N GPUs and N SSDs, you may want to have a dedicated per GPU)
    offload_params: bool = False
```

# Parts of Implementation
 1) Efficient Chunked Disk I/O functions
 2) Tensor Subclass implementation that exposes Tensors that can migrate back and forth between CPU and disk (SsdTensorHandle and SsdParameter). This also includes some optimizations for pickling (torch.save-ing) SsdTensorHandles (and its descendants) allowing them to serialize to disk and back w/out having multiple entire copies of the tensor in memory at a time.
 3) Implementation of Flattened Parameter (SsdFlatParameter/SsdFlatParameterView) and ancillary code to support them. Current implementation uses non-chainable derivative of pytorch parameterization I've coined propertization (turning an nn.Module attribute into a property -- necessary for updating views from the flattened tensor due to .data modification) I'm also looking into another possible alternative that would utilize a transparently wrapped SsdFlatParameterView object (similar to wrapt, but much simpler).

### Additional context

# Future Works/Desires
* Creating Asynchronous disk I/O calls that can be called by an prefetcher/offloader to pipeline disk accesses.
* Simplifying SsdTensorHandle by removing .data overriding within FSDP
* Benchmarking + Profiling both time and memory usage of ssd_offloaded tensors when used in realistic models.
   * Optimizations to improve performance. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by zhaojuanmao** [comment]
for the API, looks like it can just add one more field "offload_directory" in current "CPUOffload" config? "offload_params" can be shared for both CPU offload and future SSD offload

**Comment by rohan-varma** [comment]
Thanks for putting this proposal together @another-pjohnson! High level ideas look good to me.

1. Re: API, I am trying to understand if we want to just add `offload_directory` but keep the name as CPU offload, this seems to lose the clarity that it is SSD offload? @zhaojuanmao. Of course, we need to be wary about backwards compatibility

I like the idea of a `OffloadConfig`, with proper documentation:

```
class OffloadConfig:
    offload_type: OffloadType.CPU or OffloadType.SSD or None
    offload_directory: only for OffloadType.SSD
```

regarding `offload_params` flag, I wonder if we need to even keep this, as there isn't really anything else to offload (offloading params implies grads are also offloaded, and it doesn't seem to make sense to only offload gradients but not parameters, which will require manual moving of grads / params in the optimizer). We could consider whether we want to CPU offload buffers or not, but I don't expect this to be a huge use case.

2. Could we have some ideas around potential use cases? I'm assuming this is to increase model scale when even CPU offload won't cut it. What sort of workloads / scale will SSD aim to enable? 

3. Would be good to add any details around how tensor subclassing would work for `SSDTensorHandle` / `SSDParameter`. 
4. Will we ensure this works with all FSDP features out of the box (backward prefetch, different sharding config, mixed precision, etc)?

**Comment by another-pjohnson** [comment]
1) `offload_params` seems to act as the "enable flag" for the CPUOffload option:

e.g.:
```
if self.cpu_offload.offload_params
...
```
I believe if we have an offload_type with CPU, SSD, and None, the above code could be replaced with:
```
if self.offload.offload_type == OffloadType.CPU:
...
elif self.offload.offload_type == OffloadType.SSD:
...
else: # offload_type == OffloadType.None (or just None)
```
This completely removes the need for offload_params as you mentioned.

Also, should `OffloadConfig.offload_type/offload_directory` be shortened to `OffloadConfig.type/directory`?

Assuming we are storing the OffloadConfig as an attribute `self.offload` it gets a little redundant to type `self.offload.offload_type` rather than `self.offload.type`.

2) use-cases would be running models that are larger than can fit on the GPU + CPU memory. Examples of this are very large transformer models. See Microsoft's [Zero Infinity](https://arxiv.org/abs/2104.07857) arxiv paper. It also allows studying or fine-tuning of models that were trained on large clusters of GPUs on a much smaller cluster. (For example research university wants to play with GPT-X type models, but only has a 64 GPU cluster.)

I would say the most useful cases to use ssd offload are on very large dense networks where their compute time is long. Since in FSDP all local shards can be loaded in parallel, the effective bandwidth for loading parameters from ssd scales linearly. With enough ssds, and long computation time, the latency of reading a parameter from ssd can be hidden by prefetching it a one layer early.

3) We will have one example of overriding `SsdTensorHandle` / `SsdParameter` because we will add `SsdFlatParameter` which is derived from `SsdParameter` and adds in the capabilities of torch.distributed.fsdp.flatten_params_wrapper:FlatParameter. If you could give some more specific details on what you're looking for, I could definitely make sure the in-code documentation covers that.

4) That will be the plan to support all FSDP features out of the box. If any features are not supported by cpu offload yet, they will probably not be supported by ssd offload. As part of the PRs we can make sure all test cases work with both ssd offload and cpu offload.

**Comment by mori360** [comment]
Solved by the PR.
